vscode-test related updates

octref opened the pull request:
See context at https://github.com/microsoft/vscode/issues/71048.

@joaomoreno Please take a look at the release notes, testing/CI docs and see if it makes sense for you. I think it looks simpler.

joaomoreno commented:
Adding @bpasero for review, since he was the original author of this.

joaomoreno commented on api/working-with-extensions/continuous-integration.md:
```diff
@@ -9,108 +9,58 @@ MetaDescription: Use Continuous Integration for testing Visual Studio Code exten
-| `CODE_LOCALE`             | Display language to use when running the tests (default is English)                            |
-| `CODE_DISABLE_EXTENSIONS` | Disable all other extensions except the one that is being tested                               |
-| `CODE_TESTS_DATA_DIR`     | Allows to specify the user-data-dir for the tests to use and thus enables to run multiple tests at the same time |
+Extension tests can be run on CI services. The `vscode-test` repository itself contains a sample extension that is tested on Azure Devops Pipelines. You can check out the [build pipeline](https://dev.azure.com/vscode/VSCode/_build?definitionId=14) or jump directly to the [build definition yaml file](https://github.com/microsoft/vscode-test/blob/master/azure-pipelines.yml).
```
The link https://github.com/microsoft/vscode-test/blob/master/azure-pipelines.yml doesn't work.

joaomoreno commented on api/working-with-extensions/continuous-integration.md:
```diff
@@ -9,108 +9,58 @@ MetaDescription: Use Continuous Integration for testing Visual Studio Code exten
-| `CODE_LOCALE`             | Display language to use when running the tests (default is English)                            |
-| `CODE_DISABLE_EXTENSIONS` | Disable all other extensions except the one that is being tested                               |
-| `CODE_TESTS_DATA_DIR`     | Allows to specify the user-data-dir for the tests to use and thus enables to run multiple tests at the same time |
+Extension tests can be run on CI services. The `vscode-test` repository itself contains a sample extension that is tested on Azure Devops Pipelines. You can check out the [build pipeline](https://dev.azure.com/vscode/VSCode/_build?definitionId=14) or jump directly to the [build definition yaml file](https://github.com/microsoft/vscode-test/blob/master/azure-pipelines.yml).
```
I've created a different public Azure DevOps project for this: https://dev.azure.com/vscode/vscode-test. Let's move the `microsoft.vscode-test` pipeline out of the VSCode project.

octref replied:
Here's what I get:

![image](https://user-images.githubusercontent.com/4033249/58350603-3599cf00-7e1b-11e9-99d6-8acb4252a4c7.png)

joaomoreno replied:
Updated. Give it another try.

joaomoreno commented on api/working-with-extensions/continuous-integration.md:
```diff
@@ -9,108 +9,58 @@ MetaDescription: Use Continuous Integration for testing Visual Studio Code exten
+  displayName: Start xvfb
+
+- bash: |
+    yarn install
```
Simply `yarn`. Or better yet: `yarn && yarn compile && yarn test`.

joaomoreno commented on api/working-with-extensions/continuous-integration.md:
```diff
@@ -9,108 +9,58 @@ MetaDescription: Use Continuous Integration for testing Visual Studio Code exten
+    versionSpec: '8.x'
+  displayName: 'Install Node.js'
+
+- bash: |
```
Does `bash` work on Windows? Shouldn't we be using the `script` block instead?

octref replied:
https://docs.microsoft.com/en-us/azure/devops/pipelines/scripts/cross-platform-scripting?view=azure-devops&tabs=yaml#consider-bash

joaomoreno commented on api/working-with-extensions/continuous-integration.md:
```diff
@@ -9,108 +9,58 @@ MetaDescription: Use Continuous Integration for testing Visual Studio Code exten
+  displayName: 'Install Node.js'
+
+- bash: |
+    if [ $AGENT_OS == "Linux" ]; then
```
We should probably use a [`condition`](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/conditions?view=azure-devops&tabs=yaml) instead of scripting it. Something like this:

```
condition: and(succeeded(), eq(variables['Agent.OS'], 'Linux'))
```

joaomoreno commented on api/working-with-extensions/testing-extension.md:
```diff
@@ -9,87 +9,194 @@ MetaDescription: Write tests for your Visual Studio Code extension (plug-in).
+
+import { runTests } from 'vscode-test';
+
+async function go() {
```
Maybe call it `main`.

joaomoreno commented on api/working-with-extensions/testing-extension.md:
```diff
@@ -9,87 +9,194 @@ MetaDescription: Write tests for your Visual Studio Code extension (plug-in).
-![test output](images/testing-extension/test-output.png)
+Although `code` is available on the CLI, it is not available in other environments such as a Continuous Integration machine. Besides, we might want to test against a different version of VS Code. Let us first setup a script to download VS Code and run the extension test using `vscode-test`:
+
+- `npm install -D vscode-test`
```
Maybe `--save-dev` is better known.

joaomoreno commented on api/working-with-extensions/testing-extension.md:
```diff
@@ -9,87 +9,194 @@ MetaDescription: Write tests for your Visual Studio Code extension (plug-in).
+      // The path to test runner
+      // Passed to --extensionTestsPath
+      testRunnerPath
+    });
```
I'd move the comments up to the `const` declarations, so as to keep this `await` simpler. Something like this:

```ts
    // The folder containing the Extension Manifest package.json
    // Passed to `--extensionDevelopmentPath`
    const extensionPath = path.resolve(__dirname, '../../');

    // The path to test runner
    // Passed to --extensionTestsPath
    const testRunnerPath = path.resolve(__dirname, './testRunner.js');

     // Download VS Code, unzip it and run the integration test
    await runTests({ extensionPath, testRunnerPath });
```

joaomoreno commented on api/working-with-extensions/testing-extension.md:
```diff
@@ -9,87 +9,194 @@ MetaDescription: Write tests for your Visual Studio Code extension (plug-in).
+When running the extension integration test, `--extensionTestsPath` points to a **test runner**. In this sample, we explain how to setup a test runner using [Mocha](https://mochajs.org/).
 
-## Launch tests configuration
+- `npm install -D mocha @types/mocha glob @types/glob`
```
Same: `--save-dev`

joaomoreno commented on api/working-with-extensions/testing-extension.md:
```diff
@@ -9,87 +9,194 @@ MetaDescription: Write tests for your Visual Studio Code extension (plug-in).
+
+    try {
+      // Run the mocha test
+      mocha.run(failures => {
```
Simpler:

```ts
mocha.run(failures => cb(null, failures));
```

joaomoreno commented on release-notes/v1_35.md:
```diff
@@ -24,6 +24,29 @@ We really appreciate people taking a look at our new features as soon as they ar
+
+### Splitting `vscode` package into `@types/vscode` and vscode-test
+
+During the `event-stream` incident last year, we found that `vscode` package was affected as it pulls in 223 transitive dependencies totaling to 16MB. From time to time, these dependencies also cause GitHub security alerts for all extensions. Since then, we started slimming down `vscode`.
```
Maybe link this to https://code.visualstudio.com/blogs/2018/11/26/event-stream

joaomoreno commented on release-notes/v1_35.md:
```diff
@@ -24,6 +24,29 @@ We really appreciate people taking a look at our new features as soon as they ar
+- Pull `vscode.d.ts` for extension development
+- Run integration test by downloading and launching a local copy of VS Code
+
+So we splitted `vscode` into `@types/vscode` and `vscode-test`, two packages with more focused functionality.
```
s/splitted/split/

octref commented:
>  assume that yo code will still setup everything needed to run integration tests for the extension, so in most cases a user will not have to deal with vscode-test directly

Good point. I realized most people would either use `yo code` or clone the sample instead of doing the manual setup, so I:
- Added a paragraph to explain how to run integration test from `yo code` output
- Modified the rest of the document from "here's how you setup your own test script & test runner" to "here's the explanation of test script & test runner in the context of helloworld-test-sample".
